op-proposer/proposer/driver.go

```diff
@@ -269,18 +279,68 @@ func (l *L2OutputSubmitter) FetchL2OOOutput(ctx context.Context) (*eth.OutputRes
 	return output, true, nil
 }
 
-// FetchDGFOutput gets the next output proposal for the DGF.
+// FetchDGFOutput queries the DGF for the latest game and infers whether it is time to make another proposal
+// If necessary, it gets the next output proposal for the DGF, and returns it along with
+// a boolean for whether the proposal should be submitted at all.
 // The passed context is expected to be a lifecycle context. A network timeout
 // context will be derived from it.
-func (l *L2OutputSubmitter) FetchDGFOutput(ctx context.Context) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(ctx, l.Cfg.NetworkTimeout)
+func (l *L2OutputSubmitter) FetchDGFOutput(ctx context.Context) (*eth.OutputResponse, bool, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.Cfg.NetworkTimeout)
 	defer cancel()
+	callOpts := &bind.CallOpts{
+		From:    l.Txmgr.From(),
+		Context: cCtx,
+	}
 
-	blockNum, err := l.FetchCurrentBlockNumber(ctx)
+	gameCount, err := l.dgfContract.GameCount(callOpts)
 	if err != nil {
-		return nil, err
+		l.Log.Warn("Could not query DisputeGameFactory.gameCount()", "err", err)
+		return nil, false, err
+	}
+
+	if gameCount.Cmp(common.Big0) != 0 {
+		// If there is at least one game, ensure
+		// enough time has lapsed that we need to make another proposal.
+
+		latestGameIndex := new(big.Int).Sub(gameCount, big.NewInt(1))
+		latestGame, err := l.dgfContract.GameAtIndex(callOpts, latestGameIndex)
+		if err != nil {
+			l.Log.Warn(fmt.Sprintf(
+				"Could not query DisputeGameFactory.GameAtIndex(%d)",
+				latestGameIndex.Int64()), "err", err)
+			return nil, false, err
+		}
+
+		timeSinceLastGame := time.Since(time.Unix(int64(latestGame.Timestamp), 0))
+		if timeSinceLastGame <= l.Cfg.ProposalInterval {
+			l.Log.Info("Duration since last game not past proposal interval", "duration", timeSinceLastGame)
+			return nil, false, nil
+		}
+
+		l.Log.Info("Duration since last game past proposal interval, submitting proposal now", "duration", timeSinceLastGame)
+	} else {
+		// If there is not yet any game, wait one interval
+		// and then proceed with making a proposal.
+		// This is to prevent the submitter attempting to
+		// submit too soon after the AnchorStateRegistry is
+		// deployed (mostly only an issue in tests).
+		time.Sleep(l.Cfg.ProposalInterval)
+	}
```

Comment:
Because the submitter is now quite "eager", it can start proposing too early and trigger this revert on L1:

https://github.com/ethereum-optimism/optimism/blob/320bd1a720d782868e64bf8c51e63a0b72b496cf/packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol#L202-L204

Because the proposer uses an op-service txmgr, it will retry up to 30 times, waiting 2 seconds after each retry. This means the e2e test will be delayed by about 1 minute, which causes the test to time out.

Reply:
oh nasty

Reply:
Could we alternatively `return nil, false, nil` here? Then we'd also wait another poll interval, but have the polling logic isolated to the loop. We could then just `Info`-log here and return.

Reply:
If we did that, we would never make the first proposal -- we need to wait and then definitely send, we can't just wait for the condition to change because it won't change without making a proposal.

Reply:
ah gotcha!